File: src/learning_tf_py/learning_tf_py/dynamic_tf_broadcaster.py
import rclpy
from rclpy.node import Node
from tf2_ros import TransformBroadcaster
from turtlesim.msg import Pose
from geometry_msgs.msg import TransformStamped
import tf_transformations

class DynamicTfBroadcaster(Node):
    def __init__(self):
        super().__init__("dynamic_tf_broadcaster_py")
        self.get_logger().info("动态坐标广播方创建成功!")
        self.broadcaster_ = TransformBroadcaster(self)
        self.subscriber_ = self.create_subscription(
            Pose, "/turtle1/pose", self.turtle_pose_callback, 10
        )

    def turtle_pose_callback(self, pose: Pose):
        qtn = tf_transformations.quaternion_from_euler(
            0.0, 0.0, pose.theta
        )

        transform = TransformStamped()
        transform.header.stamp = self.get_clock().now().to_msg()
        transform.header.frame_id = "world"
        transform.child_frame_id = "turtle1"
        transform.transform.translation.x = pose.x
        transform.transform.translation.y = pose.y
        transform.transform.translation.z = 0.0
        transform.transform.rotation.x = qtn[0]
        transform.transform.rotation.y = qtn[1]
        transform.transform.rotation.z = qtn[2]
        transform.transform.rotation.w = qtn[3]
        
        self.broadcaster_.sendTransform(transform)

def main():
    rclpy.init()
    node = DynamicTfBroadcaster()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()

if __name__ == "__main__":
    main()

File: src/learning_tf_cpp/src/dynamic_tf_broadcaster.cpp
#include "rclcpp/rclcpp.hpp"
#include "tf2_ros/transform_broadcaster.hpp"
#include "turtlesim/msg/pose.hpp"
#include "geometry_msgs/msg/transform_stamped.hpp"
#include "tf2/LinearMath/Quaternion.hpp"

using turtlesim::msg::Pose;
using std::placeholders::_1;

class DynamicTfBroadcaster : public rclcpp::Node {
public:
    DynamicTfBroadcaster() : Node("dynamic_tf_broadcaster_cpp") {
        RCLCPP_INFO(this->get_logger(), "动态坐标广播方创建成功!");
        broadcaster_ = std::make_shared<tf2_ros::TransformBroadcaster>(this);
        subscriber_ = this->create_subscription<Pose>(
            "/turtle1/pose", 10,
            std::bind(&DynamicTfBroadcaster::turtle_pose_callback, this, _1)
        );
    }

private:
    void turtle_pose_callback(const Pose::SharedPtr pose_msg) {
        auto qtn = tf2::Quaternion();
        qtn.setRPY(0.0, 0.0, pose_msg->theta);

        auto transform = geometry_msgs::msg::TransformStamped();
        transform.header.stamp = this->now();
        transform.header.frame_id = "world";
        transform.child_frame_id = "turtle1";
        transform.transform.translation.x = pose_msg->x;
        transform.transform.translation.y = pose_msg->y;
        transform.transform.translation.z = 0.0;
        transform.transform.rotation.x = qtn.getX();
        transform.transform.rotation.y = qtn.getY();
        transform.transform.rotation.z = qtn.getZ();
        transform.transform.rotation.w = qtn.getW();

        /*
            void sendTransform(const geometry_msgs::msg::TransformStamped &transform)
        */
        broadcaster_->sendTransform(transform);
    }

    std::shared_ptr<tf2_ros::TransformBroadcaster> broadcaster_;
    rclcpp::Subscription<Pose>::SharedPtr subscriber_;
};

int main(int argc, char **argv) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<DynamicTfBroadcaster>();
    rclcpp::spin(node);
    rclcpp::shutdown();

    return 0;
}